### PR TITLE
[zero] add stateful tensor

### PR DESCRIPTION
**Pull request by feifeibear**
implement PatrickStar stateful tensor.
Towards a dynamic CPU-GPU layout of param fp16 and grad fp16 data.

### CODE REVIEW

File: colossalai/zero/sharded_param/sharded_tensor.py [hunk 1]
```diff
@@ -1,30 +1,33 @@
...
-        self.local_rank = dist.get_rank(self.process_group)
-        self._is_sharded = False
+        super().__init__(tensor)
+        self.trans_state(TensorState.HOLD)
 
         self._origin_shape = tensor.shape
         self._origin_numel = tensor.numel()
         self._origin_dtype = tensor.dtype
 
+        self.process_group = process_group
```
Comment (ver217) [inline colossalai/zero/sharded_param/sharded_tensor.py:20]: Does `ShardedTensor` still need process group now?

File: colossalai/zero/sharded_model/_utils.py [hunk 2]
```diff
@@ -35,7 +37,10 @@ def cast_tensor_to_fp16(tensor: torch.Tensor) -> torch.Tensor:
     return tensor
 
 
-def cast_tensor_to_fp32(tensor: torch.Tensor) -> torch.Tensor:
```
Comment (ver217) [inline colossalai/zero/sharded_model/_utils.py:38]: Should we also modify `cast_tensor_fo_fp16()` for stateful tensor?

File: colossalai/zero/sharded_param/sharded_tensor.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/zero/sharded_param/sharded_tensor.py:20]: DONE! removed the following 3 lines of code.

File: colossalai/zero/sharded_model/_utils.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/zero/sharded_model/_utils.py:38]: DONE

**Review by ver217 — APPROVED** [review]

**Pull request merged**